File: bindings/go/tests/behavior_tests/list_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package opendal_test

import (
	"context"
	"fmt"
	"slices"
	"strings"

	opendal "github.com/apache/opendal/bindings/go"
	"github.com/google/uuid"
	"github.com/stretchr/testify/require"
)

func testsList(cap *opendal.Capability) []behaviorTest {
	if !cap.Read() || !cap.Write() || !cap.List() || !cap.CreateDir() {
		return nil
	}
	tests := []behaviorTest{
		testListCheck,
		testListDir,
		testListRichDir,
		testListEmptyDir,
		testListNonExistDir,
		testListSubDir,
		testListNestedDir,
		testListDirWithFilePath,
		testListEntryMetadata,
		testListWithDefaultOptions,
	}
	if cap.ListWithRecursive() {
		tests = append(tests, testListWithRecursive)
	}
	if cap.ListWithLimit() {
		tests = append(tests, testListWithLimit)
	}
	if cap.ListWithStartAfter() {
		tests = append(tests, testListWithStartAfter)
	}
	if isCapEnabled(cap.ListWithVersions, "list_with_versions") {
		tests = append(tests, testListWithVersions)
	}
	if isCapEnabled(cap.ListWithDeleted, "list_with_deleted") {
		tests = append(tests, testListWithDeleted)
	}
	return tests
}

func testListCheck(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	assert.Nil(op.Check(context.Background()), "operator check must succeed")
}

func testListDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	path, content, size := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))

	assert.Nil(op.Write(context.Background(), path, content), "write must succeed")

	obs, err := op.List(context.Background(), parent)
	assert.Nil(err)
	defer obs.Close()

	var found bool
	for obs.Next() {
		entry := obs.Entry()

		if entry.Path() != path {
			continue
		}

		meta, err := op.Stat(context.Background(), entry.Path())
		assert.Nil(err)
		assert.True(meta.IsFile())
		assert.Equal(uint64(size), meta.ContentLength())
		found = true
		break
	}
	assert.Nil(obs.Error())
	assert.True(found, "file must be found in list")
}

func testListRichDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	assert.Nil(op.CreateDir(context.Background(), parent))

	var expected []string
	for range 10 {
		path, content, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))
		expected = append(expected, path)
		assert.Nil(op.Write(context.Background(), path, content))
	}

	obs, err := op.List(context.Background(), parent)
	assert.Nil(err)
	defer obs.Close()
	var actual []string
	for obs.Next() {
		entry := obs.Entry()
		actual = append(actual, entry.Path())
	}
	assert.Nil(obs.Error())

	expected = append(expected, parent)
	slices.Sort(expected)
	slices.Sort(actual)

	assert.Equal(expected, actual)
}

func testListEmptyDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	dir := fixture.NewDirPath()

	assert.Nil(op.CreateDir(context.Background(), dir), "create must succeed")

	obs, err := op.List(context.Background(), dir)
	assert.Nil(err)
	defer obs.Close()
	var paths []string
	for obs.Next() {
		entry := obs.Entry()
		paths = append(paths, entry.Path())
		assert.Equal(dir, entry.Path())
	}
	assert.Nil(obs.Error())
	assert.Equal(1, len(paths), "dir should only return itself")

	paths = nil
	obs, err = op.List(context.Background(), strings.TrimSuffix(dir, "/"))
	assert.Nil(err)
	defer obs.Close()
	for obs.Next() {
		entry := obs.Entry()
		path := entry.Path()
		paths = append(paths, path)
		meta, err := op.Stat(context.Background(), path)
		assert.Nil(err, "given dir should exist")
		assert.True(meta.IsDir(), "given dir must be dir, but found: %v", path)
	}
	assert.Nil(obs.Error())
	assert.Equal(1, len(paths), "only return the dir itself, but found: %v", paths)
}

func testListNonExistDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	dir := fixture.NewDirPath()

	obs, err := op.List(context.Background(), dir)
	assert.Nil(err)
	defer obs.Close()
	assert.False(obs.Next(), "dir should only return empty")
}

func testListSubDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	path := fixture.NewDirPath()

	assert.Nil(op.CreateDir(context.Background(), path), "create must succeed")

	obs, err := op.List(context.Background(), "/")
	assert.Nil(err)
	defer obs.Close()

	var found bool
	for obs.Next() {
		entry := obs.Entry()
		if path != entry.Path() {
			continue
		}
		found = true
		break
	}
	assert.Nil(obs.Error())
	assert.True(found, "dir should be found in list")
}

func testListNestedDir(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	dir := fixture.PushPath(fmt.Sprintf("%s%s/", parent, uuid.NewString()))

	filePath := fixture.PushPath(fmt.Sprintf("%s%s", dir, uuid.NewString()))
	dirPath := fixture.PushPath(fmt.Sprintf("%s%s/", dir, uuid.NewString()))

	assert.Nil(op.CreateDir(context.Background(), parent), "create must succeed")
	assert.Nil(op.CreateDir(context.Background(), dir), "create must succeed")
	assert.Nil(op.Write(context.Background(), filePath, []byte("test_list_nested_dir")), "write must succeed")
	assert.Nil(op.CreateDir(context.Background(), dirPath), "create must succeed")

	obs, err := op.List(context.Background(), parent)
	assert.Nil(err)
	defer obs.Close()
	var paths []string
	var foundParent bool
	var foundDir bool
	for obs.Next() {
		entry := obs.Entry()
		paths = append(paths, entry.Path())
		if entry.Path() == parent {
			foundParent = true
		} else if entry.Path() == dir {
			foundDir = true
		}
	}
	assert.Nil(obs.Error())
	assert.Equal(2, len(paths), "parent should only got 2 entry")
	assert.Equal(foundParent, true, "parent should be found in list")
	assert.Equal(foundDir, true, "dir should be found in list")

	obs, err = op.List(context.Background(), dir)
	assert.Nil(err)
	defer obs.Close()
	paths = nil
	var foundFile bool
	var foundDirPath bool
	foundDir = false
	for obs.Next() {
		entry := obs.Entry()
		paths = append(paths, entry.Path())
		if entry.Path() == filePath {
			foundFile = true
		} else if entry.Path() == dirPath {
			foundDirPath = true
		} else if entry.Path() == dir {
			foundDir = true
		}
	}
	assert.Nil(obs.Error())
	assert.Equal(3, len(paths), "dir should only got 3 entries")

	assert.True(foundDir, "dir should be found in list")

	assert.True(foundFile, "file should be found in list")
	meta, err := op.Stat(context.Background(), filePath)
	assert.Nil(err)
	assert.True(meta.IsFile())
	assert.Equal(uint64(20), meta.ContentLength())

	assert.True(foundDirPath, "dir path should be found in list")
	meta, err = op.Stat(context.Background(), dirPath)
	assert.Nil(err)
	assert.True(meta.IsDir())
}

func testListEntryMetadata(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	assert.Nil(op.CreateDir(context.Background(), parent))

	filePath, content, size := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))
	assert.Nil(op.Write(context.Background(), filePath, content))

	obs, err := op.List(context.Background(), parent)
	assert.Nil(err)
	defer obs.Close()

	var sawFile, sawDir bool
	for obs.Next() {
		entry := obs.Entry()
		meta := entry.Metadata()
		assert.NotNil(meta, "list entry metadata must always be populated for path %s", entry.Path())

		switch entry.Path() {
		case filePath:
			sawFile = true
			assert.True(meta.IsFile(), "expected file for %s", entry.Path())
			assert.False(meta.IsDir(), "expected file for %s", entry.Path())
			assert.Equal(uint64(size), meta.ContentLength(),
				"content length from list metadata must match written file size")
		case parent:
			sawDir = true
			assert.True(meta.IsDir(), "expected dir for %s", entry.Path())
			assert.False(meta.IsFile(), "expected dir for %s", entry.Path())
		}
	}
	assert.Nil(obs.Error())
	assert.True(sawFile, "file entry %s must be returned by list", filePath)
	assert.True(sawDir, "parent dir entry %s must be returned by list", parent)
}

func testListDirWithFilePath(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	path, content, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))

	assert.Nil(op.Write(context.Background(), path, content))

	obs, err := op.List(context.Background(), strings.TrimSuffix(parent, "/"))
	assert.Nil(err)
	defer obs.Close()

	for obs.Next() {
		entry := obs.Entry()
		assert.Equal(parent, entry.Path())
	}
	assert.Nil(obs.Error())
}

func testListWithDefaultOptions(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	subDir := fixture.PushPath(fmt.Sprintf("%s%s/", parent, uuid.NewString()))
	fileInParent, content, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))
	fileInSub := fixture.PushPath(fmt.Sprintf("%s%s", subDir, uuid.NewString()))

	assert.Nil(op.CreateDir(context.Background(), parent))
	assert.Nil(op.CreateDir(context.Background(), subDir))
	assert.Nil(op.Write(context.Background(), fileInParent, content))
	assert.Nil(op.Write(context.Background(), fileInSub, content))

	obs, err := op.List(context.Background(), parent)
	assert.Nil(err)
	defer obs.Close()

	var paths []string
	for obs.Next() {
		paths = append(paths, obs.Entry().Path())
	}
	assert.Nil(obs.Error())

	assert.NotContains(paths, fileInSub,
		"List without options must not descend into sub-directories")
	assert.Contains(paths, fileInParent, "direct child file must appear")
	assert.Contains(paths, subDir, "direct child dir must appear")
}

func testListWithRecursive(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	subDir := fixture.PushPath(fmt.Sprintf("%s%s/", parent, uuid.NewString()))
	deepDir := fixture.PushPath(fmt.Sprintf("%s%s/", subDir, uuid.NewString()))

	fileTop := fixture.PushPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))
	fileMid := fixture.PushPath(fmt.Sprintf("%s%s", subDir, uuid.NewString()))
	fileDeep := fixture.PushPath(fmt.Sprintf("%s%s", deepDir, uuid.NewString()))

	content := []byte("recursive test content")

	assert.Nil(op.CreateDir(context.Background(), parent))
	assert.Nil(op.CreateDir(context.Background(), subDir))
	assert.Nil(op.CreateDir(context.Background(), deepDir))
	assert.Nil(op.Write(context.Background(), fileTop, content))
	assert.Nil(op.Write(context.Background(), fileMid, content))
	assert.Nil(op.Write(context.Background(), fileDeep, content))

	obs, err := op.List(context.Background(), parent, opendal.ListWithRecursive(true))
	assert.Nil(err)
	defer obs.Close()

	var paths []string
	for obs.Next() {
		paths = append(paths, obs.Entry().Path())
	}
	assert.Nil(obs.Error())

	assert.Contains(paths, fileTop, "recursive list must include top-level file")
	assert.Contains(paths, fileMid, "recursive list must include file in sub-directory")
	assert.Contains(paths, fileDeep, "recursive list must include file in deep directory")
}

func testListWithLimit(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	assert.Nil(op.CreateDir(context.Background(), parent))

	// Write 5 files.
	for range 5 {
		path, content, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))
		assert.Nil(op.Write(context.Background(), path, content))
	}

	// List with limit=2; the operation must succeed without error.
	obs, err := op.List(context.Background(), parent, opendal.ListWithLimit(2))
	assert.Nil(err)
	defer obs.Close()

	var paths []string
	for obs.Next() {
		paths = append(paths, obs.Entry().Path())
	}
	assert.Nil(obs.Error())
	// At least one entry must be returned (limit is a hint, not a hard cap for all backends).
	assert.NotEmpty(paths, "list with limit must return at least one entry")
}

func testListWithStartAfter(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	assert.Nil(op.CreateDir(context.Background(), parent))

	// Write files with predictable sorted names.
	var filePaths []string
	for i := range 5 {
		name := fmt.Sprintf("%sfile-%02d", parent, i)
		filePaths = append(filePaths, name)
		assert.Nil(op.Write(context.Background(), name, []byte("content")))
	}
	slices.Sort(filePaths)

	// Start listing from the second file (index 1).
	pivotName := strings.TrimPrefix(filePaths[1], "/")
	obs, err := op.List(context.Background(), parent, opendal.ListWithStartAfter(pivotName))
	assert.Nil(err)
	defer obs.Close()

	var paths []string
	for obs.Next() {
		paths = append(paths, obs.Entry().Path())
	}
	assert.Nil(obs.Error())

	// Files after the pivot must appear.
	for _, p := range filePaths[2:] {
		assert.Contains(paths, p, "start_after must include entries after pivot")
	}
}

func testListWithVersions(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	path, _, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))

	assert.Nil(op.Write(context.Background(), path, []byte("version-1")), "first write must succeed")
	assert.Nil(op.Write(context.Background(), path, []byte("version-2")), "second write must succeed")

	obs, err := op.List(context.Background(), path, opendal.ListWithVersions(true))
	assert.Nil(err, "list with versions must succeed")
	defer obs.Close()

	var count int
	var currentCount int
	for obs.Next() {
		entry := obs.Entry()
		if entry.Path() == path {
			count++
			meta := entry.Metadata()
			version, ok := meta.Version()
			assert.True(ok, "version metadata must be present for list with versions")
			assert.NotEmpty(version, "each version entry must have a version ID")
			if curr, ok := meta.IsCurrent(); ok && curr {
				currentCount++
			}
		}
	}
	assert.Nil(obs.Error())
	assert.GreaterOrEqual(count, 2, "list with versions must return at least 2 entries for the same path")
	assert.Equal(1, currentCount, "exactly one version entry should be current")
}

func testListWithDeleted(assert *require.Assertions, op *opendal.Operator, fixture *fixture) {
	parent := fixture.NewDirPath()
	path, content, _ := fixture.NewFileWithPath(fmt.Sprintf("%s%s", parent, uuid.NewString()))

	assert.Nil(op.Write(context.Background(), path, content), "write must succeed")

	obs, err := op.List(context.Background(), path, opendal.ListWithDeleted(true))
	assert.Nil(err, "list with deleted must succeed before deletion")
	defer obs.Close()
	var beforeCount int
	for obs.Next() {
		if obs.Entry().Path() == path {
			beforeCount++
		}
	}
	assert.Nil(obs.Error())
	assert.Equal(1, beforeCount, "active file must appear exactly once before deletion")

	assert.Nil(op.Delete(context.Background(), path), "delete must succeed")

	obs2, err := op.List(context.Background(), path, opendal.ListWithDeleted(true))
	assert.Nil(err, "list with deleted must succeed after deletion")
	defer obs2.Close()
	var foundDeleteMarker bool
	for obs2.Next() {
		entry := obs2.Entry()
		if entry.Path() == path {
			meta := entry.Metadata()
			if meta != nil && meta.IsDeleted() {
				version, ok := meta.Version()
				assert.True(ok, "delete marker must have a version ID")
				assert.NotEmpty(version, "delete marker must have a version ID")
				foundDeleteMarker = true
				break
			}
		}
	}
	assert.Nil(obs2.Error())
	assert.True(foundDeleteMarker, "delete marker must be found after deletion")
}

File: bindings/cpp/tests/behavior/list_test.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "../framework/test_framework.hpp"
#include <set>
#include <thread>
#include <atomic>

namespace opendal::test {

class ListBehaviorTest : public OpenDALTest {
protected:
    void SetUp() override {
        OpenDALTest::SetUp();
    }
};

// Test listing empty directory
OPENDAL_TEST_F(ListBehaviorTest, ListEmptyDirectory) {
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    if (TestConfig::instance().service_name() == "alluxio") {
        GTEST_SKIP() << "Alluxio has different list behavior that includes unexpected paths";
    }
    auto dir_path = random_dir_path();
    
    // Create empty directory
    op_.CreateDir(dir_path);
    
    // List the directory
    auto entries = op_.List(dir_path);
    
    // Should contain at least the directory itself
    EXPECT_GE(entries.size(), 1);
    
    bool found_dir = false;
    for (const auto& entry : entries) {
        if (entry.path == dir_path) {
            found_dir = true;
            auto metadata = op_.Stat(entry.path);
            EXPECT_EQ(metadata.type, opendal::EntryMode::DIR);
        }
    }
    EXPECT_TRUE(found_dir);
}

// Test listing directory with files
OPENDAL_TEST_F(ListBehaviorTest, ListDirectoryWithFiles) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    auto file1_path = dir_path + "file1.txt";
    auto file2_path = dir_path + "file2.txt";
    auto file3_path = dir_path + "file3.txt";
    
    // Create directory and files
    op_.CreateDir(dir_path);
    op_.Write(file1_path, random_string(100));
    op_.Write(file2_path, random_string(200));
    op_.Write(file3_path, random_string(300));
    
    // List the directory
    auto entries = op_.List(dir_path);
    
    // Should contain directory and 3 files
    EXPECT_EQ(entries.size(), 4);
    
    std::set<std::string> expected_paths = {dir_path, file1_path, file2_path, file3_path};
    std::set<std::string> actual_paths;
    
    for (const auto& entry : entries) {
        actual_paths.insert(entry.path);
        
        auto metadata = op_.Stat(entry.path);
        if (entry.path == dir_path) {
            EXPECT_EQ(metadata.type, opendal::EntryMode::DIR);
        } else {
            EXPECT_EQ(metadata.type, opendal::EntryMode::FILE);
        }
    }
    
    EXPECT_EQ(actual_paths, expected_paths);
}

// Test listing nested directories
OPENDAL_TEST_F(ListBehaviorTest, ListNestedDirectories) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto base_dir = random_dir_path();
    auto sub_dir = base_dir + "subdir/";
    auto file1 = base_dir + "file1.txt";
    auto file2 = sub_dir + "file2.txt";
    
    // Create nested structure
    op_.CreateDir(base_dir);
    op_.CreateDir(sub_dir);
    op_.Write(file1, random_string(100));
    op_.Write(file2, random_string(200));
    
    // List base directory
    auto entries = op_.List(base_dir);
    
    std::set<std::string> expected_paths = {base_dir, sub_dir, file1};
    std::set<std::string> actual_paths;
    
    for (const auto& entry : entries) {
        actual_paths.insert(entry.path);
    }
    
    EXPECT_EQ(actual_paths, expected_paths);
    
    // List subdirectory
    auto sub_entries = op_.List(sub_dir);
    
    std::set<std::string> sub_expected_paths = {sub_dir, file2};
    std::set<std::string> sub_actual_paths;
    
    for (const auto& entry : sub_entries) {
        sub_actual_paths.insert(entry.path);
    }
    
    EXPECT_EQ(sub_actual_paths, sub_expected_paths);
}

// Test listing non-existent directory
OPENDAL_TEST_F(ListBehaviorTest, ListNonExistentDirectory) {
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    
    // Ensure directory doesn't exist
    EXPECT_FALSE(op_.Exists(dir_path));
    
    // Listing non-existent directory should either throw or return empty results
    // Different backends may handle this differently
    try {
        auto entries = op_.List(dir_path);
        // If no exception, should return empty or minimal results
        EXPECT_LE(entries.size(), 1); // At most the directory itself
    } catch (const std::exception& e) {
        // Exception is also acceptable behavior
        SUCCEED() << "Expected exception thrown: " << e.what();
    }
}

// Test listing with many files
OPENDAL_TEST_F(ListBehaviorTest, ListManyFiles) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    const int num_files = 100;
    std::vector<std::string> file_paths;
    
    // Create directory
    op_.CreateDir(dir_path);
    
    // Create many files
    for (int i = 0; i < num_files; ++i) {
        auto file_path = dir_path + "file_" + std::to_string(i) + ".txt";
        op_.Write(file_path, random_string(50));
        file_paths.push_back(file_path);
    }
    
    // List the directory
    auto entries = op_.List(dir_path);
    
    // Should contain directory + num_files
    EXPECT_EQ(entries.size(), num_files + 1);
    
    // Verify all files are present
    std::set<std::string> expected_paths(file_paths.begin(), file_paths.end());
    expected_paths.insert(dir_path);
    
    std::set<std::string> actual_paths;
    for (const auto& entry : entries) {
        actual_paths.insert(entry.path);
    }
    
    EXPECT_EQ(actual_paths, expected_paths);
}

// Test listing with special character names
OPENDAL_TEST_F(ListBehaviorTest, ListSpecialCharNames) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    auto file1 = dir_path + "file-with.special_chars.txt";
    auto file2 = dir_path + "file with spaces.txt";
    auto file3 = dir_path + "file_123.txt";
    
    // Create directory and files
    op_.CreateDir(dir_path);
    op_.Write(file1, random_string(100));
    op_.Write(file2, random_string(100));
    op_.Write(file3, random_string(100));
    
    // List the directory
    auto entries = op_.List(dir_path);
    
    std::set<std::string> expected_paths = {dir_path, file1, file2, file3};
    std::set<std::string> actual_paths;
    
    for (const auto& entry : entries) {
        actual_paths.insert(entry.path);
    }
    
    EXPECT_EQ(actual_paths, expected_paths);
}

// Test using lister iterator
OPENDAL_TEST_F(ListBehaviorTest, ListerIterator) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    auto file1_path = dir_path + "file1.txt";
    auto file2_path = dir_path + "file2.txt";
    
    // Create directory and files
    op_.CreateDir(dir_path);
    op_.Write(file1_path, random_string(100));
    op_.Write(file2_path, random_string(200));
    
    // Use lister
    auto lister = op_.List(dir_path);
    
    std::set<std::string> expected_paths = {dir_path, file1_path, file2_path};
    std::set<std::string> actual_paths;
    
    for (const auto& entry : lister) {
        actual_paths.insert(entry.path);
        
        auto metadata = op_.Stat(entry.path);
        if (entry.path == dir_path) {
            EXPECT_EQ(metadata.type, opendal::EntryMode::DIR);
        } else {
            EXPECT_EQ(metadata.type, opendal::EntryMode::FILE);
        }
    }
    
    EXPECT_EQ(actual_paths, expected_paths);
}

// Test listing root directory
OPENDAL_TEST_F(ListBehaviorTest, ListRootDirectory) {
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    // List root directory (empty path or "/")
    auto entries = op_.List("/");
    
    // Should not throw and return some entries
    EXPECT_NO_THROW({
        auto entries = op_.List("/");
    });
}

// Test metadata in list results
OPENDAL_TEST_F(ListBehaviorTest, ListMetadata) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    auto file_path = dir_path + "test_file.txt";
    auto content = random_string(1000);
    
    // Create directory and file
    op_.CreateDir(dir_path);
    op_.Write(file_path, content);
    
    // List the directory
    auto entries = op_.List(dir_path);
    
    for (const auto& entry : entries) {
        auto metadata = op_.Stat(entry.path);
        if (entry.path == file_path) {
            EXPECT_EQ(metadata.type, opendal::EntryMode::FILE);
            EXPECT_EQ(metadata.content_length, content.size());
        } else if (entry.path == dir_path) {
            EXPECT_EQ(metadata.type, opendal::EntryMode::DIR);
        }
    }
}

// Test concurrent listing
OPENDAL_TEST_F(ListBehaviorTest, ConcurrentListing) {
    OPENDAL_SKIP_IF_UNSUPPORTED_WRITE();
    OPENDAL_SKIP_IF_UNSUPPORTED_CREATE_DIR();
    OPENDAL_SKIP_IF_UNSUPPORTED_LIST();
    auto dir_path = random_dir_path();
    
    // Create directory with some files
    op_.CreateDir(dir_path);
    for (int i = 0; i < 10; ++i) {
        auto file_path = dir_path + "file_" + std::to_string(i) + ".txt";
        op_.Write(file_path, random_string(100));
    }
    
    const int num_threads = 5;
    std::vector<std::thread> threads;
    std::vector<std::vector<opendal::Entry>> results(num_threads);
    std::atomic<int> error_count{0};
    
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&, i]() {
            try {
                results[i] = op_.List(dir_path);
            } catch (const std::exception& e) {
                error_count++;
            }
        });
    }
    
    for (auto& thread : threads) {
        thread.join();
    }
    
    EXPECT_EQ(error_count, 0);
    
    // All results should be the same
    for (int i = 1; i < num_threads; ++i) {
        EXPECT_EQ(results[0].size(), results[i].size());
    }
}

} // namespace opendal::test 